### PR TITLE
[interp] Remove return value indirection

### PR DESCRIPTION
**Pull request by BrzVlad**
When returning from a method we used to save the return value into frame->retval. We would then have separate opcodes for normal and void call and, depending on the type of call, we would access the retval and push it on the stack. In case of valuetype returns, the value is not returned in the proper space so we have an additional MINT_VTRESULT that copies the return around in the caller frame. In addition to this we had retval, is_void locals, saved as part of the current frame state which just make things confusing and are not needed.

This commit simplifies the call procedure by completely moving the responsability of the return to the called method. Simply put, once a method is ready for the call, it will leave its stack and vtstack to the state after pop-ing the arguments. The called method will then directly push the result to the parent frame's stack/vtstack. This means the calling code doesn't care about the type of the return anymore, the IL instructions after the call will just expect the return on the stack, if any, and use it naturally.

We still make use of frame->retval for interp entry frames (since we don't have a parent frame to push to), for pinvoke and jit_call frames out of convenience, but we should consider removing this variable to further trim down InterpFrame.

### CODE REVIEW

File: src/mono/mono/mini/interp/interp.c [hunk 1]
```diff
@@ -3841,59 +3844,56 @@ interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClauseArgs
...
+			cmethod = get_virtual_method (cmethod, this_arg->vtable);
+			if (m_class_is_valuetype (this_arg->vtable->klass) && m_class_is_valuetype (cmethod->method->klass)) {
+				/* unbox */
+				gpointer unboxed = mono_object_unbox_internal (this_arg);
+				sp [0].data.p = unboxed;
 			}
 
 			frame->ip = ip;
 #ifdef ENABLE_EXPERIMENT_TIERED
 			ip += 5;
```
Comment (vargaz) [inline src/mono/mono/mini/interp/interp.c:3877]: This is not good anymore, not sure whenever it matters.

**Comment by vargaz** [comment]
Looks nice.

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/mini/interp/interp.c (hunk 1 above)
Comment (BrzVlad) [reply src/mono/mono/mini/interp/interp.c:3877]: Yeah, MINT_JIT_CALL2 still expects the old calling convention. Will fix it some other time to not further complicate this PR.

**Pull request merged**